### ISSUE TITLE
React DOM (e.g <div />) not type checked

### ISSUE THREAD

**Issue by aldendaniels** [facebook/flow#3773]
Flow allows this without complaint:

```typescript
import React from 'react';

class Test extends React.Component {
  render() {
    return <div style={1} bogus="allowed">Test</div>;
  }
}
```
[Try](https://flow.org/try/#0JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoctAGxQGc64AVJO+JADxiQDsATRsnQwAdAGFckHr3gBvcgEgi-JFAAUASjjy4uuEqQwArlB5wAPH2AA3OGwCe1JAF5ZARgC+cAEYQA5kZ0zgBEKNTUEADuSHzBAHwsbOYA9FbWcRR6cB7kHkA)

Should we add a disclaimer to this effect to the [Flow + React](https://flow.org/en/docs/frameworks/react/#setup-flow-with-react-a-classtoc-idtoc-setup-flow-with-react-hreftoc-setup-flow-with-reacta) until this level of type-safety is supported?

**Comment by mwalkerwells** [comment]
You mean add a disclaimer that the default return type of ```render``` is ```?React$Element<any>```?
https://github.com/facebook/flow/blob/master/lib/react.js#L32

Not sure how useful that would be. If you need more type safety, you can easily add it:
```js
import React from 'react';

class Test extends React.Component {
	render(): React$Element<{ style: number, bogus: string}> {
    	return <div style={1} bogus="allowed">Test</div>;
    }
}
```

**Comment by aldendaniels** [comment]
@mwalkerwells Yeah, that's what I mean - there are not type defs for the React primitives and it's not clear how to provide any. By contrast, Typescript [documents how to provide your own types for JSX intrinsics](https://www.typescriptlang.org/docs/handbook/jsx.html).

Regardless, I'd (incorrectly) expected types for these - and thought a disclaimer to this effect might be helpful to others. Feel free to close if you feel otherwise. Otherwise I'm happy to submit a PR.

I expect that the simplest way to get strongly typed primitives is to write wrapping components for the various DOM primitives that are strongly typed - in the spirit of [react-primitives](https://github.com/lelandrichardson/react-primitives), but sticking to DOM spec - this is something I've been looking into for use in my own projects.

**Comment by Peeja** [comment]
We would find it extremely useful to type check props on intrinsics.

That appears to be what the issue title is about, but not what @mwalkerwells's clarification points to, unless I'm misunderstanding. As a simpler example, I'd expect this to fail:

```jsx
<div style={1} bogus="allowed">Test</div>;
```

We sometimes spread props into intrinsic components, and it would be valuable to know when we've spread something that includes props React doesn't expect there, or of the wrong type.

**Comment by Peeja** [comment]
Also, it would be extra nice to have, for instance, `<button>`'s `onClick` prop automatically typed as `(SyntheticEvent<HTMLButtonElement>) => void`. We've passed handlers expecting to be attached to the wrong kind of DOM element before, and this would catch that.

**Comment by cprodescu** [comment]
@motiz88 thank you for taking this on!

**Comment by SamChou19815** [comment]
The libdef has been moved to flow-typed. If you want to make improvements, please make PR over there. We have upstreamed a much more complete version there
